src/transform.rs

```diff
@@ -12,6 +12,104 @@ extern crate libc;
 use partition::TxSize;
 use partition::TxType;
 
+use util::clamp;
+
+static COSPI_INV: [i32; 64] = [
+  4096, 4095, 4091, 4085, 4076, 4065, 4052, 4036, 4017, 3996, 3973, 3948,
+  3920, 3889, 3857, 3822, 3784, 3745, 3703, 3659, 3612, 3564, 3513, 3461,
+  3406, 3349, 3290, 3229, 3166, 3102, 3035, 2967, 2896, 2824, 2751, 2675,
+  2598, 2520, 2440, 2359, 2276, 2191, 2106, 2019, 1931, 1842, 1751, 1660,
+  1567, 1474, 1380, 1285, 1189, 1092, 995, 897, 799, 700, 601, 501, 401, 301,
+  201, 101,
+];
+
+#[inline]
+fn half_btf(w0: i32, in0: i32, w1: i32, in1: i32, bit: i32) -> i32 {
+  let result = (w0 * in0) + (w1 * in1);
+  round_shift(result, bit)
+}
+
+#[inline]
+fn round_shift(value: i32, bit: i32) -> i32 {
+  (value + (1 << (bit - 1))) >> bit
+}
+
+#[inline]
+fn clamp_value(value: i32, bit: i32) -> i32 {
+  // Do nothing for invalid clamp bit.
+  if bit <= 0 {
+    value
+  } else {
+    let max_value: i32 = ((1i64 << (bit - 1)) - 1) as i32;
+    let min_value: i32 = (-(1i64 << (bit - 1))) as i32;
+    clamp(value, min_value, max_value)
+  }
+}
+
+fn av1_idct4(input: [i32; 4], output: &mut [i32], range: i32) {
+  let cos_bit = 12;
+  let mut bf0: [i32; 4] = input;
+  let mut bf1: [i32; 4] = [0; 4];
+
+  // stage 0
+
+  // stage 1
+  bf1[0] = bf0[0];
+  bf1[1] = bf0[2];
+  bf1[2] = bf0[1];
+  bf1[3] = bf0[3];
+
+  // stage 2
+  bf0[0] = half_btf(COSPI_INV[32], bf1[0], COSPI_INV[32], bf1[1], cos_bit);
```

Comment:
This could be a single assignment.
``` rust
let bf1 = [input[0], input[2], input[1], input[3]];

let bf0 = [half_btf(COSPI_INV[32], bf1[0], COSPI_INV[32], bf1[1], cos_bit),
           half_btf(COSPI_INV[32], bf1[0], -COSPI_INV[32], bf1[1], cos_bit),
...